### ISSUE TITLE
Implement multiple currencies

### ISSUE THREAD

**Issue by Changaco** [liberapay/liberapay.com#182]
I just realized that we didn't actually have an open issue for this, so here it is.

MangoPay [currently supports](https://www.mangopay.com/features/) the following currencies in addition to euros: GBP, CAD, SEK, NOK, DKK, CHF, ZAR, PLN, USD, AUD.

[Fees](https://www.mangopay.com/pricing/pricing-worldwide/) become a lot more complicated when we introduce other currencies into the equation.

**Comment by Changaco** [comment]
I think this is still not a priority (previous statement: https://github.com/liberapay/liberapay.com/issues/44#issuecomment-159319824).

**Comment by Changaco** [comment]
I asked MangoPay if they were planning to add support for currency conversion to their API, they said no.

Not having currency conversion means that it would be more likely that a donor could find themselves in the annoying situation of having money in their Liberapay account but still needing to add more because the person they want to give to only accepts other currencies.

**Comment by MartinDelille** [comment]
Would it be possible to rely on another service for conversion?

**Comment by Changaco** [comment]
> Would it be possible to rely on another service for conversion?

Assuming that such a service exists, it would be complicated and probably not instantaneous. There would also be the problem of bypassing MangoPay's fees.

We could allow users to trade currencies between themselves, but that wouldn't guarantee always being able to convert your money when you want to.

**Comment by aggsol** [comment]
Does this include Bitcoin (XBT)?

**Comment by Changaco** [comment]
> Does this include Bitcoin (XBT)?

Not directly, but implementing native support of Bitcoins requires multiple currencies accounting. That's why #364 is tagged as Blocked on this issue.

**Comment by aggsol** [comment]
If the incoming money would always exchanged to Euro then no multiple currency accounting would be necessary. Like the the provider would accept Bitcoin, USD etc. on Liberapays behalf, exchange it to Euros.

**Comment by Changaco** [comment]
Automatic currency conversion is built into credit card systems, Liberapay donors whose bank accounts aren't denominated in euros can pay by card. This issue is about implementing multiple currencies in Liberapay itself.

**Comment by aggsol** [comment]
Just to be clear: Running everything in EUR and just displaying a converted amount in an another currency is not an option?

**Comment by Changaco** [comment]
No, we want real support for multiple currencies, it's the big missing piece of our internationalization efforts.

**Comment by Changaco** [comment]
The priority is US dollars, and we can start by only supporting one currency per account if that simplifies things.

**Comment by Changaco** [comment]
I've sent a message to Mangopay asking for clarifications on how exactly billing works for transactions in foreign currencies.

**Comment by Changaco** [comment]
One of the first steps here is to introduce currencies into the DB schema. If <https://github.com/samv/pg-currency> was still alive it would have been a nice solution.

**Comment by Changaco** [comment]
Regarding exchange rates, I guess we can use [the reference rates from the European Central Bank](http://www.ecb.europa.eu/stats/policy_and_exchange_rates/euro_reference_exchange_rates/html/index.en.html), specifically [this XML file](http://www.ecb.europa.eu/stats/eurofxref/eurofxref-daily.xml) which is updated daily around 16:00 CET.

**Comment by Changaco** [comment]
Regarding mixed-currency situations (e.g. when a European wants to donate to a US-based person), I'm thinking we could have the following user preferences to determine what to do:

- `preferred_currency`
- `accept_foreign_currencies` | Accept currencies other than `preferred_currency`? If not then the burden is on the donors, which can deter donations but simplifies things for the recipient.

As previously mentioned in <https://github.com/liberapay/liberapay.com/issues/182#issuecomment-234532120> we could also implement currency swapping, and have the following preferences:

- `convert_foreign_currencies` | Attempt to automatically swap foreign currencies with users willing to swap in the other direction?
- `trade_currencies` | Participate in currency swapping even if `convert_foreign_currencies` is turned off?

**Comment by Changaco** [comment]
WIP status: 26 commits, 83 files changed, 1291 insertions(+), 698 deletions(-).

**Comment by Changaco** [comment]
Total of the first 3 pull requests: [38 commits, 97 files changed, 2261 insertions(+), 1161 deletions(-)](https://github.com/liberapay/liberapay.com/compare/227...0b5ad1123ca5ef4834bfd446af902fbdabaf2caf).

**Comment by Changaco** [comment]
Deployment went badly, we have lots of errors, including DB self-check failures.

**Comment by Changaco** [comment]
I've managed to fix almost everything. The remaining issues aren't critical.

**Comment by Changaco** [comment]
Payments in USD weren't actually working because the currency wasn't enabled in our production account. I contacted mangopay and they've just enabled it.

**Comment by Changaco** [comment]
What should the [default currency](https://github.com/liberapay/liberapay.com/blob/228/liberapay/utils/i18n.py#L384) be once our support of the US dollar has improved and stabilized?

A relevant map: https://en.wikipedia.org/wiki/File:DOLLAR_AND_EURO_IN_THE_WORLD.svg

**Comment by mattbk** [comment]
Is this working and documented yet?

**Comment by Changaco** [comment]
The work isn't complete. I've started writing a blog post to announce the support of USD, but first I need to implement:

- [ ] changing the `main_currency` and `accept_all_currencies` settings of one's account
- [ ] proper handling of multiple currencies when adding or withdrawing money

**Comment by mattbk** [comment]
Thanks, just trying to keep up.

**Comment by Changaco** [comment]
I've been working on adapting the Giving page to multiple currencies. Here's a screenshot what I have:

![liberapay-2017-11-10-giving-page](https://user-images.githubusercontent.com/1581590/32667997-b9c82210-c63c-11e7-9c92-ff46a8d91004.png)

**Comment by Changaco** [comment]
![Screenshot of the currency settings form I've been working on](https://user-images.githubusercontent.com/1581590/33517324-ad1dc694-d782-11e7-8e68-2b3b68e7e987.png)

**Comment by Changaco** [comment]
Draft announcement: https://medium.com/@liberapay/liberapay-now-supports-us-dollars-f9484e3008f8

**Comment by Changaco** [comment]
Wondering if I should publish now or wait.

**Comment by Changaco** [comment]
Published: https://medium.com/liberapay-blog/liberapay-now-supports-us-dollars-f9484e3008f8

**Comment by Changaco** [comment]
Let's call this done and move to new issues for the rest of the work.

**Comment by Changaco** [comment]
Spanish translation of the blog post by @jorgesumle: [Liberapay ya acepta dólares americanos](https://www.freakspot.net/liberapay-ya-acepta-dolares-americanos/).
